Importer upgrades

kubark42 opened the pull request:
[Add select all/none buttons](https://github.com/d-ronin/dRonin/pull/655)
[Fix the automatic application of settinsg on import](https://github.com/d-ronin/dRonin/pull/669)

Mad kudos to @jpbarraca for fixing the apply-on-import issue, that always was a weakness of the importer.

The next step in improving the uploader would be to show the actual XML so that settings can be ported over even if the correspondence is too broken, e.g. TrimAngleSettings vs. SubTrimSettings.

jpbarraca committed:
Adds Select All/None buttons to UAV import dialog

jpbarraca committed:
Style fixes according to review

jpbarraca committed:
Updated according to comments. Save button will apply and save

jpbarraca committed:
Added: Import UAV settings without applying values automatically

peabody124 commented on ground/gcs/src/plugins/uavsettingsimportexport/importsummary.cpp:
```diff
@@ -2,6 +2,7 @@
  ******************************************************************************
  *
  * @file       importsummary.cpp
+ * @author     dRonin, http://dRonin.org/, Copyright (C) 2015
```
can you move it after TL, also since you made some changes might as well
extend TL to 2016

kubark42 replied:
I didn't make any changes. git cherry-pick flubbed because the original commit history was composed of changes in commits and changes in merges. The end result was having to merge all changes into one commit. Dunno why even though I amended the commit to put Joao as the author it wound up ultimately reverting back to me.

peabody124 commented on ground/gcs/src/plugins/uavsettingsimportexport/importsummary.cpp:
```diff
@@ -61,7 +107,7 @@ ImportSummaryDialog::~ImportSummaryDialog()
 void ImportSummaryDialog::openHelp()
 {
-    QDesktopServices::openUrl( QUrl("https://github.com/TauLabs/TauLabs/wiki/OnlineHelp:-UAV-Settings-import-export", QUrl::StrictMode) );
+    QDesktopServices::openUrl( QUrl("https://github.com/d-ronin/dRonin/wiki/OnlineHelp:-UAV-Settings-import-export", QUrl::StrictMode) );
```
This should be updated

peabody124 commented on ground/gcs/src/plugins/uavsettingsimportexport/uavsettingsimportexportfactory.h:
```diff
@@ -36,6 +39,7 @@ class UAVSETTINGSIMPORTEXPORT_EXPORT UAVSettingsImportExportFactory : public QOb
 public:
     UAVSettingsImportExportFactory(QObject *parent = 0);
     ~UAVSettingsImportExportFactory();
+    static bool updateObject(UAVObject *obj, QDomNode * node);
```
I don't see this new method being defined anywhere. Can you check if it works with this line (and the one
above on 33) deleted?

kubark42 commented on ground/gcs/src/plugins/uavsettingsimportexport/importsummary.cpp:
```diff
@@ -25,11 +26,27 @@
 #include "importsummary.h"
 
+#define UAV_SETIMPEXP_APPLY 1
+#define UAV_SETIMPEXP_SAVE 2
```
Unused variables. Remove.

kubark42 commented on ground/gcs/src/plugins/uavsettingsimportexport/importsummary.cpp:
```diff
@@ -54,6 +91,15 @@ ImportSummaryDialog::ImportSummaryDialog( QWidget *parent) :
 ImportSummaryDialog::~ImportSummaryDialog()
 {
     delete ui;
+    delete importedObjects;
```
This doesn't look like it needs deleting. In addition, there are several `new` commands in this PR, and it looks like some could in general be removed/refactored.

kubark42 replied:
Okay, tracked it down a bit further and it looks brittle. `importedObjects` is assigned locally, but it's pointer is `importedObjectManager`, which is created in `uavsettingsimportexportfactory.cpp`. So the `delete` is happening outside the originating class, and the originating class has lost track. This doesn't feel right. Gonna investigate some more.

kubark42 commented on ground/gcs/src/plugins/uavsettingsimportexport/importsummary.cpp:
```diff
@@ -37,14 +54,34 @@ ImportSummaryDialog::ImportSummaryDialog( QWidget *parent) :
+   connect(signalMapper, SIGNAL(mapped(int)), this, SLOT(doTheApplySaving(int)));
+
+   // Connect the Select All/None buttons
+   signalMapper = new QSignalMapper (this);
```
This also looks like lost memory. I don't see why not to declare these in the class.

kubark42 replied:
I've refactored all of this to be a normal signal/slot. The SignalMapping was an interesting approach to the problem of linking multiple buttons to one `SLOT`, but I prefer using `sender()` to do this.

kubark42 commented on ground/gcs/src/plugins/uavsettingsimportexport/importsummary.cpp:
```diff
@@ -25,11 +26,19 @@
  * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
  */
+
+// for XML object
```
These includes also aren't used.

kubark42 committed:
Importer: Refactor out use of QSignalMapper

Using QSignalMapper was an interesting way to disambiguate between multiple
buttons calling a single slot, but ultimately the sender() method provides
for a more compact and easy to maintain solution.

kubark42 committed:
Importer: Refacor use of pointers for importedObjectManager

There was no need to use pointer-fu for importedObjectManager. Instead,
declare it locally and pass the pointer to setUAVOSettings.

Also took the opportunity to move where the qDebug() statement fired, since
the import isn't over until the dialog box is closed.

kubark42 committed:
UAVObject importer: feature cleanup after porting

1.) Replace 'delete' architecture by standard array approach
2.) Remove extraneous header file modifications.
3.) Change URL back to Tau Labs
4.) Remove unneeded includes

kubark42 commented:
@peabody124, refactored the PR. Give it another look now.

@jpbarraca, you might give a look at the changes I added. They're another way to accomplish the same thing, but I found the structure a bit cleaner this way. What do you think?

jpbarraca commented:
@kubark42 You are correct regarding the cleaner approach. This PR evolved from a previous where it was possible to do Apply or Save independently.

kubark42 commented:
Yeah, I think it's valuable to be able to "try-before-you-buy" so prefer to have both save and apply buttons. The settings importer is pretty opaque, but your upgrade is super valuable.
